test/integration/test_components_error.py

```diff
@@ -139,6 +139,26 @@ def package_info(self):
            "to other internal components that are not defined: ['other', 'another']" in t.out
 
 
+def test_missing_external_components():
+    foo = textwrap.dedent("""
+        from conan import ConanFile
+        class Recipe(ConanFile):
+            name = "consumer"
+            version = "1.0"
+            requires = "foo/1.0", "bar/1.0"
+            def package_info(self):
+                self.cpp_info.components["comp1"].libs = []
+                self.cpp_info.components["comp2"].requires = ["comp1"]
```

Comment:
If you have an internal requirement, but no external requirements:

* The `check_component_requires` function won't trigger any warnings, because it checks if you have external dependencies before checking if you're missing any
* You won't depend on anything in say CMakeDeps, as you already define a requirement

But if I add only `foo::foo`, _then_ a complaint for the missing `bar::bar` is issued

Reply:
The current logic when no external ``requires`` are defined is that you depend on all of them. Every component will link both with ``foo::foo`` and ``bar::bar``. 
The moment that an external ``requires`` is defined, then, Conan needs to know how the other dependency is used, by which component.

So this behavior is more or less expected.
A different thing is if we want to relax the second case to a warning instead of an error, because there have been some exceptional cases in the past (like ``test_requires``, but I think that was solved) that it could give false positives. The current logic was added to avoid later obscure link errors way more difficult to understand and solve.